kliff/transforms/configuration_transforms/graphs/generate_graph.py

```diff
@@ -0,0 +1,218 @@
+from loguru import logger
+from monty.dev import requires
+
+from kliff.utils import torch_available, torch_geometric_available
+
+if torch_available():
+    import torch
+
+from typing import TYPE_CHECKING
+
+from kliff.transforms.configuration_transforms.graphs import graph_module
+
+from ..configuration_transform import ConfigurationTransform
+
+if TYPE_CHECKING:
+    from kliff.dataset import Configuration
+
+if torch_geometric_available():
+    from torch_geometric.data import Data
+
+
+if torch_geometric_available():
+
+    class PyGGraph(Data):
+        """
+        A Pytorch Geometric compatible graph representation of a configuration. When loaded
+        into a class:`torch_geometric.data.DataLoader` the graphs of type PyGGraph
+        will be automatically collated and batched.
+
+        """
+
+        def __init__(self):
+            super().__init__()
+            self.num_nodes = (
+                None  # Simplify sizes and frees up pos key word, coords is cleaner
+            )
+            self.energy = None
+            self.forces = None
+            self.n_layers = None
+            self.coords = None
+            self.images = None
+            self.species = None
+            self.z = None
+            self.contributions = None
+
+        def __inc__(self, key, value, *args, **kwargs):
+            if "index" in key or "face" in key:
+                return self.num_nodes
+            elif "contributions" in key:
+                return 2
+            elif "images" in key:
+                return torch.max(value) + 1
+            else:
+                return 0
+
+        def __cat_dim__(self, key, value, *args, **kwargs):
+            if "index" in key or "face" in key:
+                return 1
+            else:
+                return 0
+
+else:
+
+    class PyGGraph:
```

Comment:
What is the use case of this class? 

It will not work with PyG, nor with KIMDriver.

Reply:
What do mean it will not work with PyG? It is direct child of `pyg.data.Data` class, it is supposed to work with PyG dataloaders, such that it will collate the graphs into a single graph.

Reply:
The one under `if`, i.e. `class PyGGraph(Data)` makes sense. The one after `else:` is not derived from pyg Data. There is no collate_fn as well. So, I'd expect it will not work. That's why I was asking the question.

Reply:
The `else` PyGData class is just a container for the data that can still work as in graph models. It is just that it will not collate directly. Hence user will get warning once. Other option would include conditional import of classes which we are slowly moving away from. Calling collate function on this should result in error as it will not have requisite dunder functions. But for single config cases, this or the torch geometric graph will work identically.

Reply:
I'd suggest we remove this. First, two classes with the same name but different behaviors typically is a source of confusion and error. Second, we can add PyG as a dependence. It is just a python package and easy to install.

Reply:
PyG is available from the conda-forge channel. https://github.com/conda-forge/pytorch_geometric-feedstock

Reply:
OK. Would remove it and submit new commit.

Reply:
So KLIFF will have Pytorch geometric as hard dependency?

Reply:
Probably not. 

My main concern is that people who are only interested in physics-based models won't need all these. So, ideally, PyTorch and PyG should only be optional. We just issue run time errors using `@requires` and give detailed installation guides. I think that would be it. Honestly, I don't think there are many people in this category.

Reply:
Sure, no issues. Would remove the `else` `PyGGraph` defination and use `@requires`